### DISCUSSION TITLE
Having trouble eager loading polymorphic resources from an outer join - Is this a bug?

### DISCUSSION THREAD

**Discussion by jgadling** [sqlalchemy/sqlalchemy#6972] [Usage Questions] [answered]
I have an application that's making heavy use of SA polymorphic identities / joined table inheritance. At a high level, we keep track of a variety of different kinds of compute jobs, and their respective inputs and outputs in our db. Our different compute job classes/tables all inherit from a base job class, and job inputs/outputs all inherit from a base resource class (outputs from one job can be inputs to another).

I'm trying to construct a query for a specific type of resource that was created by a specific type of job, and optionally a specific type of inputs to that job, if they exist. I can get SQLAlchemy to construct the joins correctly, up until I try to eagerly load the inputs, at which point SA produces a query that creates cartesian product with my input table.

The correct join, produced when I don't add a contains_eager() for the input resources:
```SQL
SELECT ...
/* The output resource I'm trying to fetch */
FROM base_resource JOIN output_resources ON base_resource.id = output_resources.base_resource_id 
/* The specific compute job that produced that resource */
JOIN (base_compute_job JOIN specific_compute_job ON base_compute_job.id = specific_compute_job.base_compute_job_id) ON base_compute_job.id = base_resource.producing_job_id 
/* Optionally, a specific type of input to that job */
LEFT OUTER JOIN (base_resource AS base_resource_1 JOIN input_resource AS input_resource_1 ON base_resource_1.id = input_resource_1.base_resource_id) ON base_compute_job.id = base_resource_1.producing_job_id
```

But as soon as I try to eagerly load the inputs, the following join gets created:
```SQL
SELECT ...
FROM input_resource,  /* Cartesian product!!! What's up? */
base_resource JOIN output_resources ON base_resource.id = output_resources.base_resource_id 
JOIN (base_compute_job JOIN specific_compute_job ON base_compute_job.id = specific_compute_job.base_compute_job_id) ON base_compute_job.id = base_resource.producing_job_id LEFT OUTER JOIN (base_resource AS base_resource_1 JOIN input_resource AS input_resource_1 ON base_resource_1.id = input_resource_1.base_resource_id) ON base_compute_job.id = base_resource_1.producing_job_id
```

This is a representation of the relevant parts of our schema as a script that reproduces this behavior:
```python
from sqlalchemy import Column, ForeignKey, MetaData, Table, create_engine
from sqlalchemy import types
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import declarative_base, declared_attr
from sqlalchemy import event

meta = MetaData(
    naming_convention={
        "ix": "ix_%(column_0_label)s",
        "uq": "uq_%(table_name)s_%(column_0_name)s",
        "ck": "ck_%(table_name)s_%(constraint_name)s",
        "fk": "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
        "pk": "pk_%(table_name)s",
    },
)


class BaseMixin:
    id = Column(types.Integer, primary_key=True, autoincrement=True)


Base = declarative_base(cls=BaseMixin, metadata=meta)

ComputeJobInputs = Table(
    "compute_job_inputs",
    Base.metadata,
    Column("compute_job_id", ForeignKey("base_compute_job.id"), primary_key=True),
    Column("base_resource_id", ForeignKey("base_resource.id"), primary_key=True),
)


class BaseResource(Base):
    __tablename__ = "base_resource"
    resource_type = Column(types.String(50))

    __mapper_args__ = {"polymorphic_on": resource_type}

    producing_job_id = Column(types.Integer, ForeignKey("base_compute_job.id"))
    label = Column(types.String(50))
    producing_job = relationship(
        "BaseComputeJob", backref=backref("outputs", uselist=True)
    )


class BaseComputeJob(Base):
    __tablename__ = "base_compute_job"
    job_type = Column(types.String(50))

    __mapper_args__ = {"polymorphic_on": job_type}

    job_inputs = relationship(
        BaseResource,
        secondary=ComputeJobInputs,
        backref="consuming_job",
        uselist=True,
    )


class SpecificComputeJob(BaseComputeJob):
    __tablename__ = "specific_compute_job"
    __mapper_args__ = {"polymorphic_identity": "specific_compute_job"}
    base_compute_job_id = Column(
        types.Integer, ForeignKey(BaseComputeJob.id), primary_key=True
    )
    label = Column(types.String(50))


class OutputResource(BaseResource):
    __tablename__ = "output_resources"
    __mapper_args__ = {"polymorphic_identity": "OUTPUT_RESOURCE"}
    base_resource_id = Column(
        types.Integer, ForeignKey(BaseResource.id), primary_key=True
    )
    output = Column(types.String(50))


class InputResource(BaseResource):
    __tablename__ = "input_resource"
    __mapper_args__ = {"polymorphic_identity": "INPUT_RESOURCE"}
    base_resource_id = Column(types.Integer, ForeignKey(BaseResource.id), primary_key=True)
    inputdata = Column(types.String(50))


e = create_engine("sqlite:///:memory:", echo="debug")
Base.metadata.drop_all(e)
Base.metadata.create_all(e)
session = Session(e, autoflush=True, autocommit=False)


job = SpecificComputeJob(label="doing work")
output = OutputResource(output="success", producing_job=job)
job_input = InputResource(inputdata="job input data")
job.job_inputs.append(job_input)

session.add_all([output, job, job_input])
session.commit()

# The joins in this query are reasonable, but don't return information about job inputs
ok_joins = (
    session.query(OutputResource)
    .join(SpecificComputeJob, OutputResource.producing_job.of_type(SpecificComputeJob))
    .outerjoin(InputResource)
    .options(
        contains_eager(
            OutputResource.producing_job.of_type(SpecificComputeJob)
        )
    )
)

print(ok_joins)

# Attempting to eagerly load the job inputs produces a broken query
broken_joins = (
    session.query(OutputResource)
    .join(SpecificComputeJob, OutputResource.producing_job.of_type(SpecificComputeJob))
    .outerjoin(InputResource)
    .options(
        contains_eager(
            OutputResource.producing_job.of_type(SpecificComputeJob)
        ).contains_eager(SpecificComputeJob.job_inputs.of_type(InputResource))
    )
)

print(broken_joins)

#outputdata = ok_joins.one()
#print(outputdata.producing_job)
#for input_res in outputdata.producing_job.job_inputs:
#    print(input_res.inputdata)

session.close()
```

**Comment by zzzeek** [comment, marked as answer]
is `outerjoin(InputResource)` producing this?:

```
LEFT OUTER JOIN (base_resource AS base_resource_1 JOIN input_resource AS input_resource_1 ON base_resource_1.id = input_resource_1.base_resource_id) ON base_compute_job.id = base_resource_1.producing_job_id
```

see what's strange about that is that it's aliased.   There might be some legacy auto-aliasing thing happening in query.outerjoin() since you aren't specifying any aliasing, but that seems pretty strange to me;  what SQLAlchemy version are you using?

Overall the solution here is likely to use aliased() against InputResource so that it's unambiguous vs. OutputResource, which uses the same base table:

```py

   ia = aliased(InputResource)
    session.query(OutputResource)
    .join(SpecificComputeJob, OutputResource.producing_job.of_type(SpecificComputeJob))
    .outerjoin(ia)
    .options(
        contains_eager(
            OutputResource.producing_job.of_type(SpecificComputeJob)
        ).contains_eager(SpecificComputeJob.job_inputs.of_type(ia))
    )
```

these kinds of joins are always frightening though so let me know what sqlalchemy version that is, the auto-aliasing is surprising here.

**Reply by jgadling** [reply]
Oops, forgot to include that in the description:

```
# pip freeze | grep -i sqlalchemy
SQLAlchemy==1.4.23
```

**Reply by jgadling** [reply]
> is `outerjoin(InputResource)` producing this?:
> 
> ```
> LEFT OUTER JOIN (base_resource AS base_resource_1 JOIN input_resource AS input_resource_1 ON base_resource_1.id = input_resource_1.base_resource_id) ON base_compute_job.id = base_resource_1.producing_job_id
> ```

Yep, it is, though I'm not sure why either.

> Overall the solution here is likely to use aliased() against InputResource so that it's unambiguous vs. OutputResource, which uses the same base table:
> 
> ```python
>    ia = aliased(InputResource)
>    ...
> ```

The suggested fix produced some strange subqueries, but I modified it to look like so:
```python
inputalias = aliased(InputResource)
fixed_joins = (
    session.query(OutputResource)
    .join(SpecificComputeJob, OutputResource.producing_job.of_type(SpecificComputeJob))
    .outerjoin(inputalias, SpecificComputeJob.job_inputs.of_type(inputalias))
    .options(
        contains_eager(
            OutputResource.producing_job.of_type(SpecificComputeJob)
        ).contains_eager(SpecificComputeJob.job_inputs.of_type(inputalias))
    )
)
```

And the resulting query is:
```SQL
SELECT ...
FROM base_resource JOIN output_resources ON base_resource.id = output_resources.base_resource_id 
JOIN (base_compute_job JOIN specific_compute_job ON base_compute_job.id = specific_compute_job.base_compute_job_id) ON base_compute_job.id = base_resource.producing_job_id
LEFT OUTER JOIN (compute_job_inputs AS compute_job_inputs_1 
  JOIN (
    SELECT base_resource.id AS base_resource_id, base_resource.resource_type AS base_resource_resource_type, base_resource.producing_job_id AS base_resource_producing_job_id, base_resource.label AS base_resource_label, input_resource.base_resource_id AS input_resource_base_resource_id, input_resource.inputdata AS input_resource_inputdata
  FROM base_resource JOIN input_resource ON base_resource.id = input_resource.base_resource_id
  ) AS anon_1 ON anon_1.base_resource_id = compute_job_inputs_1.base_resource_id) ON base_compute_job.id = compute_job_inputs_1.compute_job_id
```

I believe this query to be essentially correct!

However, it is possible (and not all that complicated) to construct this query in SQL without the subquery. 
- Does SQLAlchemy always resolve aliased ORM classes as subqueries?
- Is there a way to construct such a query with SA?

**Reply by zzzeek** [reply]
> > is `outerjoin(InputResource)` producing this?:
> > ```
> > LEFT OUTER JOIN (base_resource AS base_resource_1 JOIN input_resource AS input_resource_1 ON base_resource_1.id = input_resource_1.base_resource_id) ON base_compute_job.id = base_resource_1.producing_job_id
> > ```
> 
> Yep, it is, though I'm not sure why either.

it might be a bug as there is some legacy logic which does this but I dont want it taking effect for new-style queries (which you aren't using here) but also it shoudn't be aliasing that entity, i dont think.

> ```
> 
> 
> I believe this query to be essentially correct!
> 
> However, it is possible (and not all that complicated) to construct this query in SQL without the subquery.

yep read on

> 
>     * Does SQLAlchemy always resolve aliased ORM classes as subqueries?

for joined table inheritance, or other mappings against joins, the default behavior is that it does.  this has historical background in the fact that some databases had problems with right-nested joins.   however, you can turn this off by passing `flat=True` to aliased(), which will give you the query you want.  I left that part off to start with since the fully "subqueried" aliases are "safer" in that they are less likely to produce overlaps of FROM objects.


> 
>     * Is there a way to construct such a query with SA?


use `aliased(InputResource, flat=True)`

**Reply by jgadling** [reply]
> > > is `outerjoin(InputResource)` producing this?:
> > > ```
> > > LEFT OUTER JOIN (base_resource AS base_resource_1 JOIN input_resource AS input_resource_1 ON base_resource_1.id = input_resource_1.base_resource_id) ON base_compute_job.id = base_resource_1.producing_job_id
> > > ```
> > 
> > 
> > Yep, it is, though I'm not sure why either.
> 
> it might be a bug as there is some legacy logic which does this but I dont want it taking effect for new-style queries (which you aren't using here) but also it shouldn't be aliasing that entity, i dont think.

Can you expand on this? What are new-style queries (should I be using them?). 

Basically, we're using joined-table inheritance very heavily, and I'm worried that subtle (possible) bugs like this or the unclear/unexpected need to use aliasing for some types of queries is going to make our application error-prone or difficult to maintain. Is this inheritance model not recommended for production use?

> use `aliased(InputResource, flat=True)`

w00! this works perfectly, thanks!

**Reply by zzzeek** [reply]
ive added #6974 illustrating the patterns that I see we are still automatically creating aliases, which are then not accommodated apparently here by contains_eager().   the mapping you are using is entirely supported, the issue is limited to when you join to an entity where there is "overlap" in the base tables, and i will have this emit a deprecation warning because the goal was that all automatic decisions like this would be made explicit.    the overlap condition will continue to be detected and will eventually (2.1) emit an error message with directions on how to resolve.

**Reply by CaselIT** [reply]
> Can you expand on this? What are new-style queries (should I be using them?).

You can start here https://docs.sqlalchemy.org/en/14/changelog/migration_14.html
To see a complete 1.4 tutorial instead go to https://docs.sqlalchemy.org/en/14/tutorial/index.html

**Reply by jgadling** [reply]
Awesome, thanks for the background!

**Reply by silenius** [reply]
I had a similar problem but with the "Row-Limited Relationships with Window Functions" pattern as specified in https://docs.sqlalchemy.org/en/14/orm/join_conditions.html?highlight=row%20limited#row-limited-relationships-with-window-functions
In my case orm.contains_eager was always broken unless I was using the partition as an alias, for ex: orm.contains_eager(foo.bar, alias=partition). Using orm.contains_eager(foo.bar.of_type(partition)) didn't worked.
